zookeeper-api/src/main/java/org/apache/helix/zookeeper/util/HttpRoutingDataReader.java

```diff
@@ -0,0 +1,134 @@
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+
+import com.jcabi.aspects.RetryOnFailure;
+import org.apache.helix.msdcommon.constant.MetadataStoreRoutingConstants;
+import org.apache.helix.msdcommon.datamodel.MetadataStoreRoutingData;
+import org.apache.helix.msdcommon.datamodel.TrieRoutingData;
+import org.apache.helix.msdcommon.exception.InvalidRoutingDataException;
+import org.apache.http.HttpEntity;
+import org.apache.http.client.methods.CloseableHttpResponse;
+import org.apache.http.client.methods.HttpGet;
+import org.apache.http.impl.client.CloseableHttpClient;
+import org.apache.http.impl.client.HttpClients;
+import org.apache.http.util.EntityUtils;
+import org.codehaus.jackson.map.ObjectMapper;
+
+
+public class HttpRoutingDataReader {
+  private static final ObjectMapper OBJECT_MAPPER = new ObjectMapper();
+  private static volatile Map<String, List<String>> _rawRoutingData;
+  private static volatile MetadataStoreRoutingData _metadataStoreRoutingData;
+
+  /**
+   * This class is a Singleton.
+   */
+  private HttpRoutingDataReader() {
+  }
+
+  /**
+   * Fetches routing data from the data source via HTTP.
+   * @return a mapping from "metadata store realm addresses" to lists of
+   * "metadata store sharding keys", where the sharding keys in a value list all route to
+   * the realm address in the key disallows a meaningful mapping to be returned
+   */
+  @RetryOnFailure(attempts = 3, delay = 1, unit = TimeUnit.SECONDS)
+  public static Map<String, List<String>> getRawRoutingData()
+      throws IOException {
+    if (_rawRoutingData == null) {
+      synchronized (HttpRoutingDataReader.class) {
+        if (_rawRoutingData == null) {
+          String msdsEndpoint =
+              System.getProperty(MetadataStoreRoutingConstants.MSDS_SERVER_ENDPOINT_KEY);
+          if (msdsEndpoint == null || msdsEndpoint.isEmpty()) {
+            throw new IllegalStateException(
+                "HttpRoutingDataReader was unable to find a valid MSDS endpoint String in System Properties!");
+          }
+
+          // Note: HttpClient's timeout settings are system timeout settings by default
+          CloseableHttpClient httpClient = HttpClients.createDefault();
```

Comment:
- `@RetryOnFailure` is applied to this whole method. It seems we are going to call this reader each time we create a RealmAwareZkClient? The heavy work is one the first time we call to get routing data. 
I would use HttpClients' retry so we don't have to acquire lock and create HttpClient 3 times (a bit heavier) if attempting. Instead, http client could just retry on exception within the locking block, which I think is easier. 
What do you think?
- When is reader supposed to be created, each time when creating zkClient, or a singleton is maintained ?

Reply:
Let me refactor the code here so that httpClient gets closed properly.

HttpRoutingDataReader is a singleton and the data is read lazily when it's called the first time, hence the double checked locking.